packages/waku/src/lib/middleware/dev-server.ts

```diff
@@ -161,13 +157,35 @@ export const devServer: Middleware = (options) => {
     }
   };
 
+  let initialModules: ClonableModuleNode[];
+
   return async (ctx, next) => {
     const [{ middleware: _removed, ...config }, vite] = await Promise.all([
       configPromise,
       vitePromise,
     ]);
+
+    // pre-process the mainJs file to see which modules are being sent to the browser by vite
+    // and using the same modules if possible in the bundlerConfig in the stream
+    const mainJs = `${config.srcDir}/${config.mainJs}`;
+    const resolved = await vite.pluginContainer.resolveId(mainJs);
+    await vite.transformRequest(resolved!.id);
+    const resolvedModule = vite.moduleGraph.idToModuleMap.get(resolved!.id)!;
+    await Promise.all(
+      [...resolvedModule.importedModules].map(({ id }) =>
+        id ? vite.warmupRequest(id) : null,
+      ),
+    );
+
+    if (!initialModules) {
```

Comment:
Do we need this `if` condition? Can we or should we compute initialModules for each request? Otherwise, should we move this before L170? Or, can we even move this logic outside of this function? `vitePromise.then(...)`

Reply:
I remember that calling in each request gave an issue that I don't remember, but let me try it again.

Reply:
The whole problem lies down to the fact that in the initial load, vite has not decided what to do for modules yet, and when things happen in parallel, it counts them as two different modules even if they are the same files.

But I wonder if that's going to cause an issue for HMR in the mainJS file.

Reply:
Can we move this before L169?